### PR TITLE
Wip/6.0 procedure

### PR DESCRIPTION
**Pull request by dreab8**

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java [hunk 1]
```diff
@@ -11,19 +11,32 @@
...
 
 /**
  * @author Steve Ebersole
  */
 public class ProcedureParameterImpl<T> extends AbstractQueryParameter<T> implements ProcedureParameterImplementor<T> {
+	private static final Logger log = Logger.getLogger( ProcedureParameterImpl.class );
+
 	private final String name;
-	private final Integer position;
+	private Integer position;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:39]: Please make this final again.

**Review by beikov — COMMENTED** [review]
Other than that, this looks good to me!

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 2]
```diff
@@ -31,81 +41,93 @@ public JdbcCall interpretCall(
...
-//		catch (SQLException e) {
-//			throw session.getJdbcServices().getSqlExceptionHelper().convert(
-//					e,
-//					"Error registering CallableStatement parameters",
-//					procedureName
-//			);
-//		}
+		// prepare parameters
+		AtomicInteger i = new AtomicInteger( 1 );
+		parameterMetadata.visitRegistrations(
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:110]: yes it was in my refactoring list :)

File: hibernate-core/src/test/java/org/hibernate/orm/test/procedure/StoredProcedureParameterTypeTest.java [hunk 3]
```diff
@@ -409,22 +407,22 @@ public void testStringTypeInParameterIsNull(SessionFactoryScope scope) {
...
-						fail("Should have thrown exception");
-					}
-					catch (IllegalArgumentException e) {
-						assertTrue( e.getMessage().endsWith( "You need to call ParameterRegistration#enablePassingNulls(true) in order to pass null parameters." ) );
-					}
-				}
-		);
-	}
+//	@Test
+//	@TestForIssue(jiraKey = "HHH-12905")
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/orm/test/procedure/StoredProcedureParameterTypeTest.java:411]: In 6 it should pass (`enablePassingNulls` is deprecated), it should only fails if we do not set the parameter value, I'm going to re-enable it and remove the  `fail("Should have thrown exception");` and add a test for when `procedureCall.setParameter` is not call

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:39]: If position is null but named parameter are not supported then I set it while registering the parameters

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+			ParameterStrategy parameterStrategy,
+			ProcedureParameterMetadataImplementor parameterMetadata,
+			SharedSessionContractImplementor session) {
+
+		final List<? extends ProcedureParameterImplementor<?>> registrations = parameterMetadata.getRegistrationsAsList();
+		for ( int i = 0; i < registrations.size(); i++ ) {
+			try {
+				registrations.get( i ).prepare( statement, i, procedureCall );
+			}
+			catch (SQLException e) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java:36]: Please move the try catch out of the loop

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 5]
```diff
@@ -28,84 +31,59 @@
...
-			ParameterStrategy parameterStrategy,
-			ParameterMetadataImplementor parameterMetadata,
-			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
+		final StringBuilder buffer;
+		if ( firstParamIsRefCursor ) {
+			buffer = new StringBuilder().append( "{? = call " );
+		}
+		else {
+			buffer = new StringBuilder().append( "{call " );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:52]: Using the perfect size estimate would be awesome :)

```suggestion
			buffer = new StringBuilder(9 + procedureName.length() + registrations.size() * 2).append( "{call " );
```

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 6]
```diff
@@ -28,84 +31,59 @@
...
-	public void registerParameters(
-			String procedureName,
-			CallableStatement statement,
-			ParameterStrategy parameterStrategy,
-			ParameterMetadataImplementor parameterMetadata,
-			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
+		final StringBuilder buffer;
+		if ( firstParamIsRefCursor ) {
+			buffer = new StringBuilder().append( "{? = call " );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:49]: Using the perfect size estimate would be awesome :)

```suggestion
			buffer = new StringBuilder(11 + procedureName.length() + registrations.size() * 2).append( "{?=call " );
```

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java [hunk 7]
```diff
@@ -47,86 +45,41 @@ public StandardCallableStatementSupport(boolean supportsRefCursors) {
 	public JdbcCall interpretCall(
 			String procedureName,
 			FunctionReturnImpl functionReturn,
-			ParameterMetadataImplementor parameterMetadata,
+			ProcedureParameterMetadataImplementor parameterMetadata,
 			ProcedureParamBindings paramBindings,
 			SharedSessionContractImplementor session) {
 		final StringBuilder buffer = new StringBuilder().append( "{call " )
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java:51]: Using the perfect size estimate would be awesome :)

```suggestion
		final StringBuilder buffer = new StringBuilder(9 + procedureName.length() + registrations.size() * 2).append( "{call " )
```

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java (hunk 4 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java]: done :)

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java (hunk 5 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java]: done, Thanks!

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java (hunk 7 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java]: done :+1:

**Comment by dreab8** [comment]
Need to fix failures for `OracleStoredProcedureTest` , `ResultMappingTest` and `StoredProcedureResultSetMappingTest`

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 8]
```diff
@@ -28,84 +31,60 @@
...
-			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
+		final List<? extends ProcedureParameterImplementor<?>> registrations = parameterMetadata.getRegistrationsAsList();
+
+		final StringBuilder buffer;
+		if ( firstParamIsRefCursor ) {
+			buffer = new StringBuilder(9 + procedureName.length() + registrations.size() * 2).append( "{? = call " );
+		}
+		else {
+			buffer = new StringBuilder(5 + procedureName.length() + registrations.size() * 2).append( "{call " );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:54]:
```suggestion
			buffer = new StringBuilder( 9 + procedureName.length() + registrations.size() * 2 ).append( "{call " );
```

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 9]
```diff
@@ -28,84 +31,60 @@
...
-			CallableStatement statement,
-			ParameterStrategy parameterStrategy,
-			ParameterMetadataImplementor parameterMetadata,
-			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
+		final List<? extends ProcedureParameterImplementor<?>> registrations = parameterMetadata.getRegistrationsAsList();
+
+		final StringBuilder buffer;
+		if ( firstParamIsRefCursor ) {
+			buffer = new StringBuilder(9 + procedureName.length() + registrations.size() * 2).append( "{? = call " );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:51]:
```suggestion
			buffer = new StringBuilder( 11 + procedureName.length() + registrations.size() * 2 ).append( "{?=call " );
```

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java [hunk 10]
```diff
@@ -47,86 +45,41 @@ public StandardCallableStatementSupport(boolean supportsRefCursors) {
 	public JdbcCall interpretCall(
 			String procedureName,
 			FunctionReturnImpl functionReturn,
-			ParameterMetadataImplementor parameterMetadata,
+			ProcedureParameterMetadataImplementor parameterMetadata,
 			ProcedureParamBindings paramBindings,
 			SharedSessionContractImplementor session) {
-		final StringBuilder buffer = new StringBuilder().append( "{call " )
+		final List<? extends ProcedureParameterImplementor<?>> registrations = parameterMetadata.getRegistrationsAsList();
+		final StringBuilder buffer = new StringBuilder(5 + procedureName.length() + registrations.size() * 2).append( "{call " )
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java:52]:
```suggestion
		final StringBuilder buffer = new StringBuilder( 9 + procedureName.length() + registrations.size() * 2 ).append( "{call " )
```

**Review by beikov — COMMENTED** [review]
Unfortunately you didn't apply the proper values for the StringBuilder capacity, so I added another comment with the correct values :)

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java (hunk 8 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:54]: why 9 and not 5 ?

File: hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingImpl.java [hunk 11]
```diff
@@ -105,12 +103,23 @@ public void setBindValue(T value) {
...
+			if ( bindType != null ) {
+				value = bindType.getExpressableJavaTypeDescriptor().coerce( value, this );
+			}
+			else if ( queryParameter.getHibernateType() != null ) {
+				value = queryParameter.getHibernateType().getExpressableJavaTypeDescriptor().coerce( value, this );
+			}
 		}
-		else if ( queryParameter.getHibernateType() != null ) {
-			value = queryParameter.getHibernateType().getExpressableJavaTypeDescriptor().coerce( value, this );
+		catch (CoercionException ce) {
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingImpl.java:114]: removed my change!

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+			ProcedureCallImplementor procedureCall,
+			CallableStatement statement,
+			ParameterStrategy parameterStrategy,
+			ProcedureParameterMetadataImplementor parameterMetadata,
+			SharedSessionContractImplementor session) {
+
+		final List<? extends ProcedureParameterImplementor<?>> registrations = parameterMetadata.getRegistrationsAsList();
+		try {
+			for ( int i = 0; i < registrations.size(); i++ ) {
+				final ProcedureParameterImplementor<?> paramater = registrations.get( i );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java:37]: Typo

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+				if ( paramater.getMode() == ParameterMode.REF_CURSOR ) {
+					if ( procedureCall.getParameterStrategy() == ParameterStrategy.NAMED ) {
+						procedureCall.getSession().getFactory().getServiceRegistry()
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/AbstractStandardCallableStatementSupport.java:40]: I think I'd prefer if you initialize `procedureCall.getSession().getFactory().getServiceRegistry().getService( RefCursorSupport.class )` once to a variable outside of the loop.

**Review by beikov — APPROVED** [review]
Just some small final comments, but then it is good to go IMO :)

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 14]
```diff
@@ -28,84 +31,60 @@
...
 			FunctionReturnImpl functionReturn,
-			ParameterMetadataImplementor parameterMetadata,
+			ProcedureParameterMetadataImplementor parameterMetadata,
 			ProcedureParamBindings paramBindings,
 			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
+		final boolean firstParamIsRefCursor = parameterMetadata.getParameterCount() != 0
+				&& isFirstParameterModeRefCursor( parameterMetadata );
 
-//		// if there are any parameters, see if the first is REF_CURSOR
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:36]: If these blocks are no longer needed, can we clean them up?

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java [hunk 15]
```diff
@@ -28,84 +31,60 @@
...
+
+		final StringBuilder buffer;
+		if ( firstParamIsRefCursor ) {
+			buffer = new StringBuilder(11 + procedureName.length() + registrations.size() * 2).append( "{?=call " );
+		}
+		else {
+			buffer = new StringBuilder(9 + procedureName.length() + registrations.size() * 2).append( "{call " );
+		}
 
-//		// prepare parameters
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/PostgresCallableStatementSupport.java:87]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureCallImpl.java [hunk 16]
```diff
@@ -432,22 +442,27 @@ private ProcedureOutputsImpl buildOutputs() {
 		//		both: (1) add the `? = ` part and also (2) register a REFCURSOR parameter for DBs (Oracle, PGSQL) that
 		//		need it.
 
-//		final String call = getSession().getJdbcServices().getJdbcEnvironment().getDialect().getCallableStatementSupport().renderCallableStatement(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureCallImpl.java:435]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java [hunk 17]
```diff
@@ -116,29 +130,29 @@ public int hashCode() {
 		return Objects.hash( name, position, mode );
 	}
 
-
 	@Override
 	public void prepare(
 			CallableStatement statement,
 			int startIndex,
 			ProcedureCallImplementor<?> procedureCall) throws SQLException {
-		throw new NotYetImplementedFor6Exception( getClass() );
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:125]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java [hunk 18]
```diff
@@ -152,88 +166,112 @@ public void prepare(
 //						);
 //					}
 //				}
-//				// TODO: sqlTypesToUse.length > 1 does not seem to have a working use case (HHH-10769).
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:155]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java [hunk 19]
```diff
@@ -152,88 +166,112 @@ public void prepare(
...
+			final JdbcTypeDescriptor recommendedJdbcType = typeToUse.getExpressableJavaTypeDescriptor()
+					.getRecommendedJdbcType( typeConfiguration.getCurrentBaseSqlTypeIndicators() );
+
+			if ( procedureCall.getParameterStrategy() == ParameterStrategy.NAMED &&
+					canDoNameParameterBinding( typeToUse, procedureCall ) ) {
+				statement.registerOutParameter( getName(), recommendedJdbcType.getJdbcTypeCode() );
+			}
+			else {
 //					for ( int i = 0; i < sqlTypesToUse.length; i++ ) {
-//						statement.registerOutParameter( startIndex + i, sqlTypesToUse[i] );
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:170]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java [hunk 20]
```diff
@@ -152,88 +166,112 @@ public void prepare(
...
-//								"Stored procedure [%s] IN/INOUT parameter [%s] not bound and `passNulls` was set to false; assuming procedure defines default value",
-//								procedureCall.getProcedureName(),
-//								this
-//						);
+				if ( position == null ) {
+					position = startIndex;
+				}
+				statement.registerOutParameter( startIndex, recommendedJdbcType.getJdbcTypeCode() );
 //					}
-//				}
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/ProcedureParameterImpl.java:209]: Another commented-out block

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java [hunk 21]
```diff
@@ -47,86 +45,41 @@ public StandardCallableStatementSupport(boolean supportsRefCursors) {
...
-	@Override
-	public void registerParameters(
-			String procedureName,
-			CallableStatement statement,
-			ParameterStrategy parameterStrategy,
-			ParameterMetadataImplementor parameterMetadata,
-			SharedSessionContractImplementor session) {
-		throw new NotYetImplementedFor6Exception( getClass() );
-
-//		final AtomicInteger count = new AtomicInteger( 1 );
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java:108]: Another commented-out block.

Plus, throws NotYetImplemented exception.  Is this still pending?

**Review by sebersole — COMMENTED** [review]
While this is not a simple "re-implement what was already supported in 5", I think we should do https://hibernate.atlassian.net/browse/HHH-13983 for 6.0.  Maybe that does not need to be done for Beta1, but it would be awesome - I wonder if there is time

File: hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java (hunk 21 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/procedure/internal/StandardCallableStatementSupport.java:108]: Ignore my "block comments" comment.  Was just being blind.  Still early ;)

**Comment by beikov** [comment]
As discussed with @sebersole, we will look into https://hibernate.atlassian.net/browse/HHH-13983 in a follow up PR, the rest of the comments was probably an oversight as the commented code was removed.

**Pull request merged**